### ISSUE TITLE
Autoscaler goes into a CLBO if we run helm upgrade to enable it (after dns setup)

### ISSUE THREAD

**Issue by prabalsharma** [cloudfoundry-incubator/kubecf#928]
**Describe the bug**
Autoscaler goes into a CLBO if we run helm upgrade to enable it on a running cluster with DNS setup.
`{"API":"https://cloud-controller-ng.service.cf.internal:9024","error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized"`

Similar to https://github.com/cloudfoundry-incubator/kubecf/issues/897

**To Reproduce**

- Deploy HA without autoscaler.
- Set up DNS ( cf login should work)
- Do a helm upgrade to turn autoscaler on.

**Expected behavior**

```
scf           asactors-0                                                        5/6     CrashLoopBackOff   26         71m
scf           asapi-0                                                           0/6     Init:6/12          2          71m
scf           asdatabase-0                                                      4/4     Running            0          71m
scf           asmetrics-0                                                       0/5     Init:5/10          2          72m
scf           asnozzle-0                                                        0/4     Init:5/9           2          70m
```

```
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine_ctl.log, [2020-05-25 21:15:33+0000] Removing stale pidfile...
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"form":{"client_id":["app_autoscaler"],"client_secret":"*REDACTED*","grant_type":["client_credentials"]},"session":"1","tokenURL":"*REDACTED*"},"log_level":1,"log_time":"2020-05-25T21:15:34Z","message":"scalingengine.cf.request-client-credential-grant","source":"scalingengine","timestamp":"1590441334.403366089"}
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized","session":"1"},"log_level":2,"log_time":"2020-05-25T21:15:34Z","message":"scalingengine.cf.request-client-credential-grant-response","source":"scalingengine","timestamp":"1590441334.771143198"}
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"API":"https://cloud-controller-ng.service.cf.internal:9024","error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized"},"log_level":2,"log_time":"2020-05-25T21:15:34Z","message":"scalingengine.failed to login cloud foundry","source":"scalingengine","timestamp":"1590441334.771569729"}
```
**Environment**
cf-operator-4.5.6+0.gffc6f942
kubecf-2.0.0-rc2
GKE
HA

**Additional context**

```
kubectl logs -n scf asactors-0 -c logs
2020/05/25 21:15:28, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"4"},"log_level":1,"log_time":"2020-05-25T21:15:24Z","message":"operator.cf.login","source":"operator","timestamp":"1590441324.428386450"}
2020/05/25 21:15:28, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"4"},"log_level":1,"log_time":"2020-05-25T21:15:24Z","message":"operator.cf.retrieve-endpoints","source":"operator","timestamp":"1590441324.429250240"}
2020/05/25 21:15:28, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"form":{"client_id":["app_autoscaler"],"client_secret":"*REDACTED*","grant_type":["client_credentials"]},"session":"4","tokenURL":"*REDACTED*"},"log_level":1,"log_time":"2020-05-25T21:15:24Z","message":"operator.cf.request-client-credential-grant","source":"operator","timestamp":"1590441324.817896843"}
2020/05/25 21:15:28, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized","session":"4"},"log_level":2,"log_time":"2020-05-25T21:15:25Z","message":"operator.cf.request-client-credential-grant-response","source":"operator","timestamp":"1590441325.155614138"}
2020/05/25 21:15:28, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"API":"https://cloud-controller-ng.service.cf.internal:9024","error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized"},"log_level":2,"log_time":"2020-05-25T21:15:25Z","message":"operator.failed to login cloud foundry","source":"operator","timestamp":"1590441325.156081676"}
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"1"},"log_level":1,"log_time":"2020-05-25T21:15:19Z","message":"scalingengine.cf.login","source":"scalingengine","timestamp":"1590441319.001336098"}
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"1"},"log_level":1,"log_time":"2020-05-25T21:15:19Z","message":"scalingengine.cf.retrieve-endpoints","source":"scalingengine","timestamp":"1590441319.001640797"}
2020/05/25 21:15:28, /var/vcap/sys/log/operator_ctl.log, [2020-05-25 21:15:22+0000] ------------ STARTING operator_ctl at Mon May 25 21:15:21 UTC 2020 --------------
2020/05/25 21:15:28, /var/vcap/sys/log/operator_ctl.err.log, [2020-05-25 21:15:22+0000] ------------ STARTING operator_ctl at Mon May 25 21:15:21 UTC 2020 --------------
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine_ctl.log, [2020-05-25 21:15:18+0000] ------------ STARTING scalingengine_ctl at Mon May 25 21:15:17 UTC 2020 --------------
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"form":{"client_id":["app_autoscaler"],"client_secret":"*REDACTED*","grant_type":["client_credentials"]},"session":"1","tokenURL":"*REDACTED*"},"log_level":1,"log_time":"2020-05-25T21:15:19Z","message":"scalingengine.cf.request-client-credential-grant","source":"scalingengine","timestamp":"1590441319.397721052"}
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine_ctl.err.log, [2020-05-25 21:15:18+0000] ------------ STARTING scalingengine_ctl at Mon May 25 21:15:17 UTC 2020 --------------
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized","session":"1"},"log_level":2,"log_time":"2020-05-25T21:15:20Z","message":"scalingengine.cf.request-client-credential-grant-response","source":"scalingengine","timestamp":"1590441320.444021940"}
2020/05/25 21:15:28, /var/vcap/sys/log/scheduler.err.log, [2020-05-25 21:15:19+0000] ------------ STARTING scheduler at Mon May 25 21:15:19 UTC 2020 --------------
2020/05/25 21:15:28, /var/vcap/sys/log/scheduler.log, [2020-05-25 21:15:19+0000] ------------ STARTING scheduler at Mon May 25 21:15:19 UTC 2020 --------------
2020/05/25 21:15:28, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"API":"https://cloud-controller-ng.service.cf.internal:9024","error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized"},"log_level":2,"log_time":"2020-05-25T21:15:20Z","message":"scalingengine.failed to login cloud foundry","source":"scalingengine","timestamp":"1590441320.444509983"}
2020/05/25 21:15:33, /var/vcap/sys/log/scalingengine_ctl.log, [2020-05-25 21:15:33+0000] ------------ STARTING scalingengine_ctl at Mon May 25 21:15:33 UTC 2020 --------------
2020/05/25 21:15:33, /var/vcap/sys/log/scalingengine_ctl.err.log, [2020-05-25 21:15:33+0000] ------------ STARTING scalingengine_ctl at Mon May 25 21:15:33 UTC 2020 --------------
2020/05/25 21:15:33, /var/vcap/sys/log/scalingengine_ctl.log, [2020-05-25 21:15:33+0000] pidno 15
2020/05/25 21:15:33, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"1"},"log_level":1,"log_time":"2020-05-25T21:15:33Z","message":"scalingengine.cf.login","source":"scalingengine","timestamp":"1590441333.760634661"}
2020/05/25 21:15:33, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"1"},"log_level":1,"log_time":"2020-05-25T21:15:33Z","message":"scalingengine.cf.retrieve-endpoints","source":"scalingengine","timestamp":"1590441333.761253357"}
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine_ctl.log, [2020-05-25 21:15:33+0000] Removing stale pidfile...
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"form":{"client_id":["app_autoscaler"],"client_secret":"*REDACTED*","grant_type":["client_credentials"]},"session":"1","tokenURL":"*REDACTED*"},"log_level":1,"log_time":"2020-05-25T21:15:34Z","message":"scalingengine.cf.request-client-credential-grant","source":"scalingengine","timestamp":"1590441334.403366089"}
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized","session":"1"},"log_level":2,"log_time":"2020-05-25T21:15:34Z","message":"scalingengine.cf.request-client-credential-grant-response","source":"scalingengine","timestamp":"1590441334.771143198"}
2020/05/25 21:15:34, /var/vcap/sys/log/scalingengine/scalingengine.stdout.log, {"data":{"API":"https://cloud-controller-ng.service.cf.internal:9024","error":"request client credential grant failed: https://uaa.prabal-gke.ci.kubecf.charmedquarks.me/oauth/token [401] 401 Unauthorized"},"log_level":2,"log_time":"2020-05-25T21:15:34Z","message":"scalingengine.failed to login cloud foundry","source":"scalingengine","timestamp":"1590441334.771569729"}
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log,   .   ____          _            __ _ _
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log,  /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log, ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log,  \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log,   '  |____| .__|_| |_|_| |_\__, | / / / /
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log,  =========|_|==============|___/=/_/_/_/
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log,  :: Spring Boot ::       (v1.5.21.RELEASE)
2020/05/25 21:15:43, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 
2020/05/25 21:15:45, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:15:45 [main] INFO  org.cloudfoundry.autoscaler.scheduler.SchedulerApplication - Starting SchedulerApplication v1.0-SNAPSHOT on asactors-0.asactors.scf.svc.cluster.local with PID 13 (/var/vcap/packages/scheduler/scheduler-1.0-SNAPSHOT.war started by root in /var/vcap/packages/scheduler)
2020/05/25 21:15:45, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:15:45 [main] INFO  org.cloudfoundry.autoscaler.scheduler.SchedulerApplication - No active profile set, falling back to default profiles: default
2020/05/25 21:15:58, /var/vcap/sys/log/scheduler/scheduler.stderr.log, WARNING: An illegal reflective access operation has occurred
2020/05/25 21:15:58, /var/vcap/sys/log/scheduler/scheduler.stderr.log, WARNING: Illegal reflective access by org.springframework.cglib.core.ReflectUtils$1 (jar:file:/var/vcap/packages/scheduler/scheduler-1.0-SNAPSHOT.war!/WEB-INF/lib/spring-core-4.3.24.RELEASE.jar!/) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
2020/05/25 21:15:58, /var/vcap/sys/log/scheduler/scheduler.stderr.log, WARNING: Please consider reporting this to the maintainers of org.springframework.cglib.core.ReflectUtils$1
2020/05/25 21:15:58, /var/vcap/sys/log/scheduler/scheduler.stderr.log, WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
2020/05/25 21:15:58, /var/vcap/sys/log/scheduler/scheduler.stderr.log, WARNING: All illegal access operations will be denied in a future release
2020/05/25 21:16:21, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:21 [localhost-startStop-1] INFO  org.springframework.web.context.ContextLoader - Root WebApplicationContext: initialization completed in 35404 ms
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/v1/apps/{app_id}/schedules],methods=[GET]}" onto public org.springframework.http.ResponseEntity<org.cloudfoundry.autoscaler.scheduler.rest.model.ApplicationSchedules> org.cloudfoundry.autoscaler.scheduler.rest.ScheduleRestController.getAllSchedules(java.lang.String)
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/v1/apps/{app_id}/schedules],methods=[PUT]}" onto public org.springframework.http.ResponseEntity<java.util.List<java.lang.String>> org.cloudfoundry.autoscaler.scheduler.rest.ScheduleRestController.createSchedules(java.lang.String,java.lang.String,org.cloudfoundry.autoscaler.scheduler.rest.model.ApplicationSchedules)
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/v1/apps/{app_id}/schedules],methods=[DELETE]}" onto public org.springframework.http.ResponseEntity<java.util.List<java.lang.String>> org.cloudfoundry.autoscaler.scheduler.rest.ScheduleRestController.deleteSchedules(java.lang.String)
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/v1/syncSchedules],methods=[PUT]}" onto public org.springframework.http.ResponseEntity<org.cloudfoundry.autoscaler.scheduler.rest.model.SynchronizeResult> org.cloudfoundry.autoscaler.scheduler.rest.ScheduleSyncRestController.synchronizeSchedules()
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/swagger-resources/configuration/ui]}" onto public org.springframework.http.ResponseEntity<springfox.documentation.swagger.web.UiConfiguration> springfox.documentation.swagger.web.ApiResourceController.uiConfiguration()
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/swagger-resources]}" onto public org.springframework.http.ResponseEntity<java.util.List<springfox.documentation.swagger.web.SwaggerResource>> springfox.documentation.swagger.web.ApiResourceController.swaggerResources()
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/swagger-resources/configuration/security]}" onto public org.springframework.http.ResponseEntity<springfox.documentation.swagger.web.SecurityConfiguration> springfox.documentation.swagger.web.ApiResourceController.securityConfiguration()
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/error]}" onto public org.springframework.http.ResponseEntity<java.util.Map<java.lang.String, java.lang.Object>> org.springframework.boot.autoconfigure.web.BasicErrorController.error(javax.servlet.http.HttpServletRequest)
2020/05/25 21:16:52, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:52 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping - Mapped "{[/error],produces=[text/html]}" onto public org.springframework.web.servlet.ModelAndView org.springframework.boot.autoconfigure.web.BasicErrorController.errorHtml(javax.servlet.http.HttpServletRequest,javax.servlet.http.HttpServletResponse)
2020/05/25 21:16:55, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:55 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter - Looking for @ControllerAdvice: org.springframework.boot.context.embedded.AnnotationConfigEmbeddedWebApplicationContext@235ecd9f: startup date [Mon May 25 21:15:45 UTC 2020]; root of context hierarchy
2020/05/25 21:16:56, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:56 [main] INFO  org.springframework.web.servlet.handler.SimpleUrlHandlerMapping - Mapped URL path [/webjars/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2020/05/25 21:16:56, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:56 [main] INFO  org.springframework.web.servlet.handler.SimpleUrlHandlerMapping - Mapped URL path [/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2020/05/25 21:16:56, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:56 [main] INFO  org.springframework.web.servlet.mvc.method.annotation.ExceptionHandlerExceptionResolver - Detected @ExceptionHandler methods in controllerExceptionHandler
2020/05/25 21:16:56, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:16:56 [main] INFO  org.springframework.web.servlet.handler.SimpleUrlHandlerMapping - Mapped URL path [/**/favicon.ico] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2020/05/25 21:17:02, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:17:02 [main] INFO  org.springframework.web.servlet.DispatcherServlet - FrameworkServlet 'dispatcherServlet': initialization started
2020/05/25 21:17:08, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:17:08 [main] INFO  org.springframework.web.servlet.DispatcherServlet - FrameworkServlet 'dispatcherServlet': initialization completed in 6140 ms
2020/05/25 21:17:09, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:17:08 [main] INFO  org.cloudfoundry.autoscaler.scheduler.SchedulerApplication$$EnhancerBySpringCGLIB$$a788f585 - Scheduler is ready to start
2020/05/25 21:17:09, /var/vcap/sys/log/scheduler/scheduler.stdout.log, 2020-05-25 21:17:08 [main] INFO  org.cloudfoundry.autoscaler.scheduler.SchedulerApplication - Started SchedulerApplication in 89.536 seconds (JVM running for 109.168)
2020/05/25 21:31:38, /var/vcap/sys/log/operator_ctl.log, [2020-05-25 21:26:30+0000][2020-05-25 21:31:38+0000] ------------ STARTING operator_ctl at Mon May 25 21:31:38 UTC 2020 --------------
2020/05/25 21:31:38, /var/vcap/sys/log/operator_ctl.log, [2020-05-25 21:31:38+0000] pidno 13
2020/05/25 21:31:38, /var/vcap/sys/log/operator_ctl.err.log, [2020-05-25 21:31:38+0000] ------------ STARTING operator_ctl at Mon May 25 21:31:38 UTC 2020 --------------
2020/05/25 21:31:38, /var/vcap/sys/log/operator_ctl.log, [2020-05-25 21:31:38+0000] Removing stale pidfile...
2020/05/25 21:31:38, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"4"},"log_level":1,"log_time":"2020-05-25T21:31:38Z","message":"operator.cf.login","source":"operator","timestamp":"1590442298.859424591"}
2020/05/25 21:31:38, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"infoURL":"https://cloud-controller-ng.service.cf.internal:9024/v2/info","session":"4"},"log_level":1,"log_time":"2020-05-25T21:31:38Z","message":"operator.cf.retrieve-endpoints","source":"operator","timestamp":"1590442298.873975992"}
2020/05/25 21:31:38, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"form":{"client_id":["app_autoscaler"],"client_secret":"*REDACTED*","grant_type":["client_credentials"]},"session":"4","tokenURL":"*REDACTED*"},"log_level":1,"log_time":"2020-05-25T21:31:38Z","message":"operator.cf.request-client-credential-grant","source":"operator","timestamp":"1590442298.900875807"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"addr":"0.0.0.0:6208","session":"20"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.health-server.new-health-server","source":"operator","timestamp":"1590442299.426009417"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"sql: no rows in result set","session":"19"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.lock-db.no-lock-found","source":"operator","timestamp":"1590442299.534282446"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"LastModifiedTimestamp":"0001-01-01T00:00:00Z","Owner":"91695a87-892f-42ca-7f77-b0dbb955b86a","Ttl":15000000000,"session":"19"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.lock-db.inserting-the-lock-details","source":"operator","timestamp":"1590442299.553835630"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"session":"19"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.lock-db.acquired-lock-successfully","source":"operator","timestamp":"1590442299.556629419"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"isLockAcquired":true,"owner":"91695a87-892f-42ca-7f77-b0dbb955b86a"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.lock-acquired-in-first-attempt","source":"operator","timestamp":"1590442299.558390856"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"refresh_interval":3600000000000,"session":"8"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.instancemetrics-dbpruner.started","source":"operator","timestamp":"1590442299.558592081"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"refresh_interval":3600000000000,"session":"10"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.appmetrics-dbpruner.started","source":"operator","timestamp":"1590442299.558894634"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"refresh_interval":3600000000000,"session":"12"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.scalingengine-dbpruner.started","source":"operator","timestamp":"1590442299.559152365"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"refresh_interval":600000000000,"session":"14"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.scalingengine-sync.started","source":"operator","timestamp":"1590442299.559481621"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"before":1590355899558786600,"error":"pq: relation \"appinstancemetrics\" does not exist","query":"DELETE FROM appinstancemetrics WHERE timestamp \u003c= $1","session":"1"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.instancemetrics-db.failed-prune-instancemetric-from-appinstancemetrics-table","source":"operator","timestamp":"1590442299.560044050"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"pq: relation \"appinstancemetrics\" does not exist","session":"7"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.instancemetrics-dbpruner.failed-prune-metrics","source":"operator","timestamp":"1590442299.560299635"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"before":1590269499559057000,"error":"pq: relation \"app_metric\" does not exist","query":"DELETE FROM app_metric WHERE timestamp \u003c= $1","session":"2"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.appmetrics-db.prune-metrics-from-app_metric-table","source":"operator","timestamp":"1590442299.560461521"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"pq: relation \"app_metric\" does not exist","session":"9"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.appmetrics-dbpruner.failed-prune-appmetrics","source":"operator","timestamp":"1590442299.560603380"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"refresh_interval":600000000000,"session":"16"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.scheduler-sync.started","source":"operator","timestamp":"1590442299.560739994"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"refresh_interval":86400000000000,"session":"18"},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.application-sync.started","source":"operator","timestamp":"1590442299.565934896"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{},"log_level":1,"log_time":"2020-05-25T21:31:39Z","message":"operator.started","source":"operator","timestamp":"1590442299.566364527"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"pq: relation \"policy_json\" does not exist","query":"SELECT app_id FROM policy_json","session":"5"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.policy-db.get-appids-from-policy-table","source":"operator","timestamp":"1590442299.566941261"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"pq: relation \"policy_json\" does not exist","session":"17"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.application-sync.failed-to-get-apps","source":"operator","timestamp":"1590442299.567125797"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://scalingengine.service.cf.internal:6104/v1/syncSchedules: dial tcp: lookup scalingengine.service.cf.internal on 10.35.240.56:53: no such host","session":"13"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.scalingengine-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590442299.614684820"}
2020/05/25 21:31:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://autoscalerscheduler.service.cf.internal:6102/v1/syncSchedules: dial tcp: lookup autoscalerscheduler.service.cf.internal on 10.35.240.56:53: no such host","session":"15"},"log_level":2,"log_time":"2020-05-25T21:31:39Z","message":"operator.scheduler-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590442299.614971638"}
2020/05/25 21:41:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://scalingengine.service.cf.internal:6104/v1/syncSchedules: dial tcp: lookup scalingengine.service.cf.internal on 10.35.240.56:53: no such host","session":"13"},"log_level":2,"log_time":"2020-05-25T21:41:39Z","message":"operator.scalingengine-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590442899.569072247"}
2020/05/25 21:41:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://autoscalerscheduler.service.cf.internal:6102/v1/syncSchedules: dial tcp: lookup autoscalerscheduler.service.cf.internal on 10.35.240.56:53: no such host","session":"15"},"log_level":2,"log_time":"2020-05-25T21:41:39Z","message":"operator.scheduler-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590442899.577484608"}
2020/05/25 21:51:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://autoscalerscheduler.service.cf.internal:6102/v1/syncSchedules: dial tcp: lookup autoscalerscheduler.service.cf.internal on 10.35.240.56:53: no such host","session":"15"},"log_level":2,"log_time":"2020-05-25T21:51:39Z","message":"operator.scheduler-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590443499.578248262"}
2020/05/25 21:51:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://scalingengine.service.cf.internal:6104/v1/syncSchedules: dial tcp: lookup scalingengine.service.cf.internal on 10.35.240.56:53: no such host","session":"13"},"log_level":2,"log_time":"2020-05-25T21:51:39Z","message":"operator.scalingengine-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590443499.580857277"}
2020/05/25 22:01:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://scalingengine.service.cf.internal:6104/v1/syncSchedules: dial tcp: lookup scalingengine.service.cf.internal on 10.35.240.56:53: no such host","session":"13"},"log_level":2,"log_time":"2020-05-25T22:01:39Z","message":"operator.scalingengine-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590444099.647038221"}
2020/05/25 22:01:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://autoscalerscheduler.service.cf.internal:6102/v1/syncSchedules: dial tcp: lookup autoscalerscheduler.service.cf.internal on 10.35.240.56:53: no such host","session":"15"},"log_level":2,"log_time":"2020-05-25T22:01:39Z","message":"operator.scheduler-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590444099.651080608"}
2020/05/25 22:11:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://scalingengine.service.cf.internal:6104/v1/syncSchedules: dial tcp: lookup scalingengine.service.cf.internal on 10.35.240.56:53: no such host","session":"13"},"log_level":2,"log_time":"2020-05-25T22:11:39Z","message":"operator.scalingengine-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590444699.569514751"}
2020/05/25 22:11:39, /var/vcap/sys/log/operator/operator.stdout.log, {"data":{"error":"Put https://autoscalerscheduler.service.cf.internal:6102/v1/syncSchedules: dial tcp: lookup autoscalerscheduler.service.cf.internal on 10.35.240.56:53: no such host","session":"15"},"log_level":2,"log_time":"2020-05-25T22:11:39Z","message":"operator.scheduler-sync.failed-to-send-sync-scheduler-request","source":"operator","timestamp":"1590444699.571097612"}
```

**Comment by prabalsharma** [comment]
Workaround is same as https://github.com/cloudfoundry-incubator/kubecf/issues/897; delete `asactors-0` pod

**Comment by andreas-kupries** [comment]
Notes:
  - Deployment via `catapult`.
  - Configuration:
    ```
    export AUTOSCALER=false
    export BACKEND=gke
    export CLUSTER_NAME=_928auto
    export DEBUG_MODE=true
    export ENABLE_EIRINI=false
    export HA=true
    export KUBECF_TEST_SUITE=brain
    export SCF_CHART= (( aws s3 cp s3://cap-release-artifacts/kubecf-2.0.0-rc2.tgz kubecf-bundle-2.0.0-rc2.tgz ))
    # Note the name change above, inserting "-bundle" before "-2.0.0...", required for catapult.
    export SCF_OPERATOR=true
    export SCF_TESTGROUP=true
    export GKE_CLUSTER_NAME=ak-kubecf-928-trial
    export GKE_CRED_JSON= (( base64 -decode the gke-private-key-base64 value in the blackbox ))
    make k8s scf scf-login
    ```
    Comes with all necessary tools (terraform, gcloud, etc.).
    DNS setup is automatic.
  - Upgrade via `helm upgrade susecf-scf (...)/kubecf/ --namespace scf --set "features.autoscaler.enabled=true" --reuse-values`

Reproduced at this point.

Note: It was necessary to wait for a long while for the UAA pods to come back before I was sure of the repro. Because the reported errors in the description are about requests to the UAA.

This also means that these may be red herrings, i.e. errors reported while `asactor` was waiting for the UAA to be back.

Currently trying to get more information out of the logs.

**Comment by andreas-kupries** [comment]
The crashing/restating containers are two, `scalingengine-scalingengine` and `operator-operator`.
Interesting: Wrt restart counts both containers had 12 restarts at the time I looked. In the short output from `get pods` this is reported as the sum, i.e. 24.

**Comment by andreas-kupries** [comment]
Workaround (asactor kill/restart) confirmed. Extended `klog` a bit to pull the rendered job files, not just logs. Review of the two pulls (before and after `asactor` restart) to come.

**Comment by gaktive** [comment]
Sounds like the same workaround, so the proposed release note for the current version still stands. Let me know if I'm wrong, then we need to let Docs know.

**Comment by andreas-kupries** [comment]
@gaktive Yes, I believe that it is the same workaround, so (IMHO) you can tell docs.

**Comment by andreas-kupries** [comment]
First review of the klog data pulled before and after pod restart.

1. Before the restart the scalingengine and operator containers were not up. That made it impossible to pull their logs and rendered scripts. 
1. However copies of the same were seen in the data from the other container (shared (eph) disk ? - confirmed)
1. Using the pulled data in direct comparison the only config difference found is the IP address assigned to the loggrator-agent pod or container. The log differences in init containers are only timestamps. For the job container they are only timestamps for the working containers. For the failing containers they are timestamps and just that they do not fail and continue to success. No direct clue towards the cause of the issue.
1. Cross comparison of operator/scaling-engine config taken from the scheduler job vs the same from o/se jobs shows no differences. So this idea that the job containers share a disk for rendered scripts and logs looks to be confirmed.

More notes
1. As the config files from the failing and restarted pods, notably the certs, __are identical__, I do not believe that the config in itself is the issue.
1. One anomaly I saw were the timestamps on logs for the failing containers themselves. Even with the containers restarting these stamps did __not change__, nor did their contents. As if during a container restart the execution of the entrypoint does not even reach a point where the logs would be modified (i.e. cleared).
1. I am now wondering if the __presence of some file__ from the first start is __preventing__ the proper restart at container level.
1. That theory would also explain why a pod-level restart can work, as this __wipes the shared disk__.
1. As a side note, this could mean that SA mode might have the same problem.

Time to look deeper into the autoscaler job scripts, and also see if there is a way to get more data out of a container start, i.e. determine how far it gets (The current regular log is abysmally short and useless). This might require patching somewhere. While doing that I will run an SA-based setup in the background to also check on the last point above.

**Comment by andreas-kupries** [comment]
Looking at the pulled pod description reminds that a container start involves the `container-run` helper/launcher application. Which got code to set a unix socket for semi-remote control. ... (Merge commit https://github.com/cloudfoundry-incubator/quarks-operator/commit/52a73efb482b65323bf8637167c016c3e5a71b00) ... Ok, based on my check of the container-run sources the socket file is re-created on each entry into the listener loop (See https://github.com/cloudfoundry-incubator/quarks-operator/blob/master/container-run/pkg/containerrun/containerrun.go#L190-L193 ). This does not fully falsify my idea that a pre-existing socket file interferes with `container-run` after container restart. It might still error in removing and this error might be swallowed, i.e. invisible. Something to look into. But for now move forward into the `ctl` scripts of the failing jobs.

**Comment by andreas-kupries** [comment]
Idea wrt patching: Side access from a working container to the shared disk, and patch the scripts used by the failing containers from there.

**Comment by andreas-kupries** [comment]
Sigh. SA mode failed, but differently - The asdatabase pod is not starting, the postgress job has a server cert which was not properly rendered:

```
% cat /var/vcap/jobs/postgres/config/certificates/server.public_cert
((postgres_server.certificate))
```

Note, watching the upgraded pods I expected things to succeed, because the uaa pod was fully active before the `asactors` pod was even through its init containers, i.e. I would have expected the actual containers to immediately talk to uaa when reached.

**Comment by andreas-kupries** [comment]
Second HA trial ... Looking at the unix socket I do see these re-created when the container restarts (per timestamps). The `RemoveAll` mentioned previously is working. That means that the restart reaches at least `container-run` itself. The time stamps on the log files are confirmed unchanged, ditto sizes.

The `.pid` files for the failing containers are not updated either, i.e. their timestamps seem to be set to the initial failure still. Either the `pid_guard` function is not reached, or it fails, i.e. thinks that there is a running process.

The `pid_guard` is responsible for the `STARTING ...` line in the logs.
And that line shows up in the `logs` container output ... Taken from the `xxx_ctl.log` files (This is not the stdout/err logs whose timestamps do not change).

I see at least some `remove stale pidfile`. That is also from the `pid_guard`.

Adding `set -x` (tracing) to the `..._ctl` scripts.

At the beginning of the utils.sh file sourced by the ctl script is
```
exec > >(tee -a >(logger -p user.info -t vcap.$(basename $0).stdout) | awk -W interactive '{ system("echo -n [$(date +\"%Y-%m-%d %H:%M:%S%z\")]"); print " " $0 }' >>/var/vcap/sys/log/$(basename $0).log)
exec 2> >(tee -a >(logger -p user.error -t vcap.$(basename $0).stderr) | awk -W interactive '{ system("echo -n [$(date +\"%Y-%m-%d %H:%M:%S%z\")]"); print " " $0 }' >>/var/vcap/sys/log/$(basename $0).err.log)
```

Tracing from that is the last output. ... After that come the function definitions, and the utils end.
No tracing from the function definitions.

Added a bit more output to be sure.

Ah. The exec's above redirect the output into the log files picked up by the `logs` container.
And somehow during the editing of the file the scaling-engine container got properly started.
It seems to be when the pid changed that things worked.

Yeah. I was editing a file, so a `vi` process was open, and that shifted the pid of the new _ctl process and the pid_guard cleaned up the stale file and exited successfully, and now the operator is started as well.

So, right now the theory is that the `pid_guard` is broken in some way, (mis)detecting in the new container that the scalingengine and operator are running and thus deciding to not restart them. And it seems to have something to do with the pids of these processes not changing from old container to new container.

At this point I have a running system, but still no proper logs because the action of setting such up raced with the restart to the point where it did not matter anymore.

So, do another repro of the issue and see if I manage to make my log changes without havign a restart interfering.

Oh. I can now say for certain that the initial issue is a race with uaa.
Basically the upgrade has a number of statefulsets restarting their pods, among them UAA.
The sequencing restarts the `uaa-1` pod first, and then the `uaa-0` pod.
It is while the `uaa-1` is still initializing that `asactors-0` completes init and starts its job containers.
Thus `operator` and `scalingengine` find the old `uaa-0` pod and when they talk it they are rejected.
It is only after that, that the `uaa-1` completes restart, then `uaa-0` completes restart, and then a pod-level restart of `asactors` has it talking to the new uaa with the matching authorization tokens, etc. whatever.

**Comment by andreas-kupries** [comment]
> Yeah. I was editing a file, so a vi process was open, and that shifted the pid of the new `_ctl` process and the `pid_guard` cleaned up the stale file and exited successfully, and now the operator is started as well.

No, that does not make sense. The `vi` process was in the `scheduler` container, that should not have any effect on the processes in `operator` or `scalingengine`.

Yet it was at that very time that things went ok.

**Comment by andreas-kupries** [comment]
Current HA trial fails with the same database error I saw for my SA trial.
(I.e. Missing rendering/templatng of some variables in certs).

```
/:/var/vcap/jobs/loggregator_agent# cat /var/vcap/jobs/postgres/config/certificates/server.*
((postgres_ca.certificate))
((postgres_server.private_key))
((postgres_server.certificate))
```

Given the time I will redo the trial tomorrow.
This issue with missing templating is something needing its own ticket.

@gaktive ^^

**Comment by andreas-kupries** [comment]
@gaktive Stupid system voodoo at work in the issue: kernel pid choices and re-mounted shared disk, and the really old-style of process guarding.

@mook-as @jandubois @viovanov Please have a read and comment with your ideas on how to fix this stupid thing.

Note, the final conclusions are at the bottom. I left some of my bad notes from the session in. I.e. some of my musings in between are wrong. No need to correct me on them.

# Session log ...

Ok. I have now a cluster running with the issue in progress, and managed to get a log out of a restart, without my editing having bumped anything to working:

```
kubectl logs -f -n scf asactors-0 -c operator-operator
+ set -e
+ JOB_DIR=/var/vcap/jobs/operator
+ RUN_DIR=/var/vcap/sys/run/operator
+ LOG_DIR=/var/vcap/sys/log/operator
+ CONF_DIR=/var/vcap/jobs/operator/config
+ HOOK_LOG_OUT=/var/vcap/sys/log/operator/hooks.stdout.log
+ HOOK_LOG_ERR=/var/vcap/sys/log/operator/hooks.stderr.log
+ PIDFILE=/var/vcap/sys/run/operator/operator.pid
+ COMMON_DIR=/var/vcap/packages/common
+ mkdir -p /var/vcap/sys/log
+ case $1 in
+ pid_guard /var/vcap/sys/run/operator/operator.pid operator
+ tee /dev/stderr
++ basename /var/vcap/jobs/operator/bin/operator_ctl
++ date
+ echo '------------ STARTING operator_ctl at Thu May 28 18:48:14 UTC 2020 --------------'
------------ STARTING operator_ctl at Thu May 28 18:48:14 UTC 2020 --------------
+ pidfile=/var/vcap/sys/run/operator/operator.pid
+ name=operator
+ '[' -f /var/vcap/sys/run/operator/operator.pid ']'
++ head -1 /var/vcap/sys/run/operator/operator.pid
------------ STARTING operator_ctl at Thu May 28 18:48:14 UTC 2020 --------------
+ pid=13
+ echo pidno 13
+ '[' -n 13 ']'
+ '[' -e /proc/13 ']'
+ echo 'operator is already running, please stop it first'
+ exit 1
pidno 13
operator is already running, please stop it first
```
What we see here is the operator container attempting a restart, and the `pid_guard` function coming to the belief that there is an operator running and thus not starting anything new. The new container exists, kube considers it a crash, and a bit later kube tries again.

The troublesome spot is the `'[' -e /proc/13 ']'` condition. The `pid_guard` reads the `.pid` file, sees that it should check for a process with id `13`, and then __finds such in the /proc filesystem__.

`/proc` is a virtual filesystem created/provided by the kernel. (`/sys` would a another).

I would think that each new container has its own instance of `/proc`. However our issue seems to indicate that the new container inherits the `/proc` of the initial failed container, and especially sees what processes the initial container was running. And thus considers these to be running in the current container as well ?!

And yesterday, when my fumbling around somehow caused the new container to fail the pid check for the main process, and the start of the new container went through fine.

And I have this again,. Had `vi` open in the scheduler container to add more logging and now the pid check in `/proc` fails, and it goes through with restart of the container, and a different pid chosen:

```
------------ STARTING operator_ctl at Thu May 28 18:58:22 UTC 2020 --------------
++ head -1 /var/vcap/sys/run/operator/operator.pid
+ pid=13
+ echo pidno 13
+ '[' -n 13 ']'
+ '[' -e /proc/13 ']'
+ echo 'Removing stale pidfile...'
+ rm /var/vcap/sys/run/operator/operator.pid
+ mkdir -p /var/vcap/sys/run/operator
pidno 13
Removing stale pidfile...
+ chown -R vcap:vcap /var/vcap/sys/run/operator
+ mkdir -p /var/vcap/sys/log/operator
+ chown -R vcap:vcap /var/vcap/sys/log/operator
+ echo 12
+ exec /var/vcap/packages/operator/operator -c /var/vcap/jobs/operator/config/operator.yml
```

# Conclusions

With more thought I now have a proper theory I am pretty sure is correct:

  1. The new container starts.
  1. The new `_ctl` process is launched by `container-run`.
  1. The kernel chooses the pid `X` for `_ctl`, which is somehow (and 99.999999% of the time) the same pid as it was chosen when the initial (failing) container was run. This means we now have `/proc/X` known in the new container.
  1. The new container mounts the same ephemeral shared disk as the initial container.
  1. As such it inherits the `/var/vcap/sys/run/*/*.pid` file.
  1. This pid file still contains `X` from the initial container.
  1. Thus the `pid_guard` comes to believe that the just-started `_ctl` process is an active already-running main process, and aborts the restart.

Fix ?

With the all the jobs in separate container and kube responsible for restarting anything, do we actually need this `pid_guard` thing anymore ?

IOW the immediate fix coming to mind is to just remove/disable the pid checks.
kube is our guard against multiple running main processes for a job.

Note: The above sequence to trouble is something I can see happening in any role, should the kernel somehow come to reuse the main process pid from a killed container in the new container for the job.
(I.e. IIRC everything has the `pid_guard` in, right ? Actually, it may have been obsoleted by the switch of upstream to BPM, and us using BPM information, but having kube as the controller).

In that case the true origin of the issue is that of auto-scaler not having switched to BPM and using the really old-style process control and guards.

**Comment by andreas-kupries** [comment]
PR with fix is #971

**Comment by mook-as** [comment]
Closing this as the relevant PR appears to have been merged.
